### PR TITLE
Add array_distinct array function

### PR DESCRIPTION
**Pull request by haozhun**
@cberner please review

### CODE REVIEW

File: presto-main/src/test/java/com/facebook/presto/type/TestArrayOperators.java [hunk 1]
```diff
@@ -281,6 +281,22 @@ public void testSort()
         }
     }
 
+    @Test void testDistinct()
+            throws Exception
+    {
+        assertFunction("ARRAY_DISTINCT(ARRAY [])", ImmutableList.of());
+
+        // Order matters here. Result should be stable.
+        assertFunction("ARRAY_DISTINCT(ARRAY [2, 3, 4, 3, 1, 2, 3])", ImmutableList.of(2L, 3L, 4L, 1L));
```
Comment (cberner) [inline presto-main/src/test/java/com/facebook/presto/type/TestArrayOperators.java]: Add tests that include NULLs

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayDistinctFunction.java [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+        if (type instanceof AbstractFixedWidthType) {
+            elementsBlockBuilder = ((AbstractFixedWidthType) type).createFixedSizeBlockBuilder(elements.size());
+        }
+        else {
+            elementsBlockBuilder = type.createBlockBuilder(new BlockBuilderStatus());
+        }
+
+        Class<?> javaType = type.getJavaType();
+        if (javaType == boolean.class) {
+            elements.stream().forEach(e -> type.writeBoolean(elementsBlockBuilder, (boolean) e));
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayDistinctFunction.java]: I don't think this is going to handle nulls correctly. Those need to be write with elementsBlockBuilder.appendNull()

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayDistinctFunction.java [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+    }
+
+    private static List<Object> blockToList(final Type type, Block block)
+    {
+        int positionCount = block.getPositionCount();
+        ArrayList<Object> result = new ArrayList<>(positionCount);
+
+        Class<?> javaType = type.getJavaType();
+        if (javaType == boolean.class) {
+            for (int i = 0; i < block.getPositionCount(); i++) {
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayDistinctFunction.java]: You need to check for nulls in block

**Comment by haozhun** [comment]
Fixed null handling and added null tests

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayDistinctFunction.java [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+        BlockBuilder elementsBlockBuilder;
+        if (type instanceof AbstractFixedWidthType) {
+            elementsBlockBuilder = ((AbstractFixedWidthType) type).createFixedSizeBlockBuilder(elements.size());
+        }
+        else {
+            elementsBlockBuilder = type.createBlockBuilder(new BlockBuilderStatus());
+        }
+
+        Class<?> javaType = type.getJavaType();
+        if (javaType == boolean.class) {
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayDistinctFunction.java]: I think this will be less code, if you switch the order of the loops and ifs, so that it's:
```
for (Object e : elements) {
if (e == null) {
appendNull()
continue;
}
if (type == ....)
}
```

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayDistinctFunction.java [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+        }
+
+        if (elements.isEmpty()) {
+            return toStackRepresentation(ImmutableList.of());
+        }
+
+        // This line of code can be removed once internal array representation becomes block
+        Block elementsBlock = listToBlock(type, elements);
+
+        Block hashBlock = TypeUtils.getHashBlock(ImmutableList.of(type), elementsBlock);
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayDistinctFunction.java]: I don't think we need this. GroupByHash will just compute the hashes for you, since you passed it Optional.empty()

**Comment by cberner** [comment]
Implementation looks right to me, except the one minor comment. @martint @electrum and @dain can you approve the function name?

**Comment by dain** [comment]
I'm fine with what ever @martint and @electrum agree on

File: presto-docs/src/main/sphinx/release/release-0.97.rst [hunk 6]
```diff
@@ -0,0 +1,8 @@
+============
+Release 0.97
+============
+
+General Changes
+---------------
+
+* Add ``array_distinct`` array function.
```
Comment (electrum) [inline presto-docs/src/main/sphinx/release/release-0.97.rst]: Remove "array" in the sentence.

Comment (electrum) [inline presto-docs/src/main/sphinx/release/release-0.97.rst]: This needs to be moved to 0.98 release notes

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayDistinctFunction.java [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+        catch (IOException e) {
+            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, e);
+        }
+
+        if (elements.isEmpty()) {
+            return toStackRepresentation(ImmutableList.of());
+        }
+
+        // This line of code can be removed once internal array representation becomes block
+        Block elementsBlock = listToBlock(type, elements);
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayDistinctFunction.java]: Should this be removed now?

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayDistinctFunction.java [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+        }
+
+        return result;
+    }
+
+    public static Slice distinct(final Type type, Slice jsonArray)
+    {
+        List<Object> elements;
+        try {
+            elements = OBJECT_MAPPER.readValue(jsonArray.getInput(), COLLECTION_TYPE);
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayDistinctFunction.java]: This needs to be updated

File: presto-main/src/test/java/com/facebook/presto/type/TestArrayOperators.java [hunk 9]
```diff
@@ -281,6 +283,26 @@ public void testSort()
...
+        assertFunction("ARRAY_DISTINCT(ARRAY [TRUE, TRUE, TRUE])", ImmutableList.of(true));
+        assertFunction("ARRAY_DISTINCT(ARRAY [TRUE, FALSE, FALSE, TRUE])", ImmutableList.of(true, false));
+        assertFunction("ARRAY_DISTINCT(ARRAY [from_unixtime(100), from_unixtime(1), from_unixtime(100)])",
+                ImmutableList.of(new SqlTimestamp(100 * 1000, TEST_SESSION.getTimeZoneKey()), new SqlTimestamp(1000, TEST_SESSION.getTimeZoneKey())));
+        assertFunction("ARRAY_DISTINCT(ARRAY ['2', '3', '2'])", ImmutableList.of("2", "3"));
+        assertFunction("ARRAY_DISTINCT(ARRAY ['BB', 'CCC', 'BB'])", ImmutableList.of("BB", "CCC"));
+
+        assertFunction("ARRAY_DISTINCT(ARRAY [NULL, 2.2, 3.3, 4.4, 3.3, 1, 2.2, 3.3])", Arrays.asList(null, 2.2, 3.3, 4.4, 1.0));
+        assertFunction("ARRAY_DISTINCT(ARRAY [2, 3, NULL, 4, 3, 1, 2, 3])", Arrays.asList(2L, 3L, null, 4L, 1L));
+        assertFunction("ARRAY_DISTINCT(ARRAY ['BB', 'CCC', 'BB', NULL])", Arrays.asList("BB", "CCC", null));
```
Comment (electrum) [inline presto-main/src/test/java/com/facebook/presto/type/TestArrayOperators.java]: Add a test for empty array and all null

File: presto-main/src/test/java/com/facebook/presto/type/TestArrayOperators.java [hunk 10]
```diff
@@ -281,6 +283,26 @@ public void testSort()
...
+    {
+        assertFunction("ARRAY_DISTINCT(ARRAY [])", ImmutableList.of());
+
+        // Order matters here. Result should be stable.
+        assertFunction("ARRAY_DISTINCT(ARRAY [2, 3, 4, 3, 1, 2, 3])", ImmutableList.of(2L, 3L, 4L, 1L));
+        assertFunction("ARRAY_DISTINCT(ARRAY [2.2, 3.3, 4.4, 3.3, 1, 2.2, 3.3])", ImmutableList.of(2.2, 3.3, 4.4, 1.0));
+        assertFunction("ARRAY_DISTINCT(ARRAY [TRUE, TRUE, TRUE])", ImmutableList.of(true));
+        assertFunction("ARRAY_DISTINCT(ARRAY [TRUE, FALSE, FALSE, TRUE])", ImmutableList.of(true, false));
+        assertFunction("ARRAY_DISTINCT(ARRAY [from_unixtime(100), from_unixtime(1), from_unixtime(100)])",
+                ImmutableList.of(new SqlTimestamp(100 * 1000, TEST_SESSION.getTimeZoneKey()), new SqlTimestamp(1000, TEST_SESSION.getTimeZoneKey())));
```
Comment (electrum) [inline presto-main/src/test/java/com/facebook/presto/type/TestArrayOperators.java]: Add a helper method `sqlTimestamp()`

**Comment by electrum** [comment]
I think this name is good. Both `distinct` and `unique` are reserved words.

**Comment by haozhun** [comment]
Updated this PR based on comments from @electrum .

File: presto-docs/src/main/sphinx/functions/array.rst [hunk 11]
```diff
@@ -32,6 +32,10 @@ Array Functions
 
     Sorts and returns the array ``x``. The elements of ``x`` must be orderable.
 
+.. function:: array_distinct(x) -> array
```
Comment (electrum) [inline presto-docs/src/main/sphinx/functions/array.rst]: Move this above `array_sort`

File: presto-docs/src/main/sphinx/release/release-0.98.rst [hunk 12]
```diff
@@ -0,0 +1,8 @@
+============
+Release 0.98
```
Comment (electrum) [inline presto-docs/src/main/sphinx/release/release-0.98.rst]: Move to next release

**Comment by electrum** [comment]
Looks good
